cudamapper/src/main.cu

```diff
@@ -116,152 +129,200 @@ int main(int argc, char* argv[])
+                                                                                                                                     overlaps_to_add));
+
+                print_pafs_futures.push_back(write_and_filter_overlaps_future);
             }
         }
-    }
 
-    // Insert empty overlap vector to denote end of processing.
-    // The lambda function for adding overlaps to queue ensures that no empty
-    // overlaps are added to the queue so as not to confuse it with the
-    // empty overlap inserted to indicate end of processing.
-    auto start_time = std::chrono::high_resolution_clock::now();
-    overlaps_writer_mtx.lock();
-    overlaps_to_write.push_back(std::vector<claragenomics::cudamapper::Overlap>());
-    overlaps_writer_mtx.unlock();
-    auto paf_time = std::chrono::duration_cast<std::chrono::milliseconds>(
-        std::chrono::high_resolution_clock::now() - start_time);
-
-    std::cerr << "\n\n"
-              << std::endl;
-    std::cerr << "Index execution time: " << index_time.count() << "ms" << std::endl;
-    std::cerr << "Matcher execution time: " << matcher_time.count() << "ms" << std::endl;
-    std::cerr << "Overlap detection execution time: " << overlapper_time.count() << "ms" << std::endl;
-    std::cerr << "PAF detection execution time: " << paf_time.count() << "ms" << std::endl;
+        // If all-to-all mapping query will no longer be needed on device, remove it from the cache
+        if (all_to_all)
+        {
+            evict_index(query_start_index, query_end_index, device_id);
+        }
 
-    // Sync overlap writer threads.
-    overlap_result.get();
+        return print_pafs_futures;
+    };
+
+    // The application (File parsing, index generation, overlap generation etc) is all launched from here.
+    // The main application works as follows:
+    // 1. Generate a thread pool with one worker thread per device (GPU).
+    // 2. For each worker, assign a query-target-range, this is list of query reads and a *list of lists of target reads* to match against.
+    // 3. Each worker returns a vector of futures (since overlap writing is dispatched to an async thread on host). All futures are waited for before the main application exits.
+
+    // 1. create thread pool to compute overlaps. One worker thread per device.
+    ThreadPool overlap_pool(num_devices);
+
+    // 2. Enqueue the query-target ranges which need to be computed, each thread returns a vector of futures for the threads it launches
+    std::vector<std::future<std::vector<std::shared_ptr<std::future<void>>>>> overlap_futures;
+    for (int i = 0; i < query_target_ranges.size(); i++)
+    {
+        // assign chunk pairs (Query, target) to device
+        auto query_target_range = query_target_ranges[i];
+        auto device_id          = i % num_devices;
+        overlap_futures.push_back(overlap_pool.enqueue(compute_overlaps, query_target_range, device_id));
```

Comment:
I'm still not sure if this works as expected. Imagine you have 2 GPUs and 2 threads in the pool and 4 query ranges.
You'd assigned them in the following way:
range_0 -> device_0
range_1 -> device_1
range_2 -> device_0
range_3 -> device_1
Threads are taking ranges in the order they entered the queue. If we have two threads that means most likely you'd have:
thread_0 -> range_0 -> device_0
thread_1 -> range_1 -> device_1
Now imagine thread_1 finishes before thread_0 and picks up the next task, which is range_2. Now you'd have:
thread_0 -> range_0 -> device_0
thread_1 -> range_2 -> device_0
meaning there are two threads working on device_0 and none on device_1.
 Am I missing some property of `ThreadPool` or something else?

Reply:
This is true,in general I am making the assumption that it only has quite a small effect, do you have any suggestions for a clean way to make a worker thread assign its work to whichever GPU is available?

Reply:
I have a different idea. How about creating equal number of threads amd GPUs and assigning each thread to a device. Then a thread whenever it's done with its current query_index could pick up the next available query_index and work on it (this pool of query_indices should obviously be implemented in a thread-safe manner).
This essentially means assigning each thread a GPU instead of assigning each query_index a GPU

Reply:
That's a good idea - I will implement it this way, then I can also remove all the mutexes around the cache since I can guarantee that there will only be one thread per device.